drivers/watchdog/wdt_sam0.c

```diff
@@ -230,7 +230,9 @@ static int wdt_sam0_feed(struct device *dev, int channel_id)
 		return -EINVAL;
 	}
 
-	WDT_REGS->CLEAR.reg = WDT_CLEAR_CLEAR_KEY_Val;
+	if (!WDT_REGS->STATUS.bit.SYNCBUSY) {
```

Comment:
Does this mean if SYNCBUSY is set then the watchdog is not fed?

Reply:
The samd21 datasheet describes the synchronization  as follows:

_Due to asynchronicity between the main clock domain and the peripheral clock domains, some registers need to be synchronized when written or read._

_When executing an operation that requires synchronization, the Synchronization Busy bit in the Status register (STATUS.SYNCBUSY) will be set immediately, and cleared when synchronization is complete._

_If an operation that requires synchronization is executed while STATUS.SYNCBUSY=1, the bus will be stalled. All operations will complete successfully, but the CPU will be stalled and interrupts will be pending as long as the bus is stalled._

The stall is about 6 peripheral clock cycles long. Since the wdt is generally clocked by a slow clock (1 KHz or 32 KHz), the stall can be a long duration (Up to 6ms in the case of a 1 KHz clock).

Generally you can just set the register and forget it when you know the register will not be touched again before the sync is complete.  In this case, we know there will be two back to back accesses because that is a requirement to feed the wdt.  So to avoid the stall, you loop on the sync busy flag.  Avoiding the bus stall allows interrupts to occur while waiting for the busy flag.  Otherwise, interrupts pend until the bus stall is complete.

Another way to do this is to check the flag first and only perform the register operation if the sync bit is clear and return a busy flag if it is not.

Reply:
@sslupsky Did you miss the loop in the code for this? Currently, it is an `if` statement.

Reply:
Apologies, it has been a while since I looked at this.

For the sam0, you only need to feed the wdt once to reset the watchdog.  The proposed change handles an edge case where the watchdog is fed before a prior feed register sync is complete.  This eliminates a blocking wait.

Reply:
The consequence is that the watchdog would not be fed for that particular register access.  The consequence of the missed feed is that the wdt would need to be fed again before the next timeout which would occur at
T <= timeout - sync time
in the future.

For example, wdt clock = 1KHz.  timeout = 8sec.  The wdt is fed at T=0.  Another feed is attempted at T=5.9ms.  The second feed is ignored because SYNCBUSY is set.  So, the wdt would need to be fed again before the next timeout at T=8s which is 8 - 0.0059 = 7.9941s from the time of the second feed.

Reply:
I did a quick test and you're right - it took ~5 s to feed the watchdog 1000 times.

re: the change, if the watchdog was recently updated then the PR silently does something different to what the user requested.  I'd rather it returns -EBUSY (or -EAGAIN?) if the feeding was skipped.

Reply:
I updated to return -EAGAIN if the feed was skipped.